# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/wasm/AsmJS.cpp
# Commit: 8ce9bbcd26e3
# Full Hash: 8ce9bbcd26e343a8d845a817cea33fd875805541
# Author: Ryan Hunt <rhunt@eqrion.net>
# Date: 2020-08-19 10:01:16
# Regressor Bug: 1561521
# File Overlap Count: 2
# Description:
#   Bug 1561521 - Generalize table/element code for future reference-types. r=lth
#   
#   wasm::Table uses wasm::TableKind to determine the element type of the table,
#   which makes it difficult to extend tables to support more reference types. This
#   commit drops TableKind in favor of directly storing the element RefType, along
# ==============================================================================

diff -r 7805ddfe2bfa -r 8ce9bbcd26e3 js/src/wasm/AsmJS.cpp
--- a/js/src/wasm/AsmJS.cpp	Tue Aug 18 16:58:32 2020 +0000
+++ b/js/src/wasm/AsmJS.cpp	Tue Aug 18 16:58:48 2020 +0000
@@ -1714,6 +1714,7 @@
     if (!seg) {
       return false;
     }
+    seg->elemType = RefType::func();
     seg->tableIndex = tableIndex;
     seg->offsetIfActive = Some(InitExpr::fromConstant(LitVal(uint32_t(0))));
     seg->elemFuncIndices = std::move(elems);
@@ -1995,7 +1996,8 @@
     }
 
     env_.asmJSSigToTableIndex[sigIndex] = env_.tables.length();
-    if (!env_.tables.emplaceBack(TableKind::AsmJS, mask + 1, Nothing())) {
+    if (!env_.tables.emplaceBack(RefType::func(), mask + 1, Nothing(),
+                                 /*isAsmJS*/ true)) {
       return false;
     }
 